File: src/statusbar/modules/battery.ts
import type {
  StatusBarModule,
  StatusBarModuleContext,
  Segment,
} from "../types.ts";

function getBatteryIcon(level: number, charging: boolean): string {
  if (charging) return "󰂄";
  if (level >= 90) return "󰁹";
  if (level >= 70) return "󰂀";
  if (level >= 50) return "󰁾";
  if (level >= 30) return "󰁼";
  if (level >= 10) return "󰁺";
  return "󰂃";
}

export const batteryModule: StatusBarModule = {
  id: "battery",
  render(ctx: StatusBarModuleContext): Segment[] {
    const bat = ctx.metrics.battery;
    if (!bat || !bat.present) return [];

    const icon = ctx.icons ? getBatteryIcon(bat.level, bat.charging) + " " : "";
    const chargingIndicator = bat.charging ? "\u26a1" : "";

    let colors;
    if (bat.level > 50) {
      colors = ctx.themeColors.modules.batteryHigh;
    } else if (bat.level > 20) {
      colors = ctx.themeColors.modules.batteryMed;
    } else {
      colors = ctx.themeColors.modules.batteryLow;
    }

    return [
      {
        text: ` ${icon}${bat.level}%${chargingIndicator} `,
        fg: colors.fg,
        bg: colors.bg,
      },
    ];
  },
};

File: rust/crates/maxmux-statusbar/src/modules/battery.rs
use crate::modules::{ModuleContext, StatusBarModule};
use crate::types::{ColorPair, Segment};

pub struct BatteryModule;

/// Choose battery icon based on level and charging state.
fn battery_icon(level: u8, charging: bool) -> &'static str {
    if charging {
        "\u{f0084}" // 󰂄
    } else if level >= 90 {
        "\u{f0079}" // 󰁹
    } else if level >= 70 {
        "\u{f0080}" // 󰂀
    } else if level >= 50 {
        "\u{f007e}" // 󰁾
    } else if level >= 30 {
        "\u{f007c}" // 󰁼
    } else if level >= 10 {
        "\u{f007a}" // 󰁺
    } else {
        "\u{f0083}" // 󰂃
    }
}

/// Choose battery color pair based on level.
pub(crate) fn battery_colors(ctx: &ModuleContext, level: u8) -> ColorPair {
    if level > 50 {
        ctx.theme_colors.modules.battery_high.clone()
    } else if level >= 20 {
        ctx.theme_colors.modules.battery_med.clone()
    } else {
        ctx.theme_colors.modules.battery_low.clone()
    }
}

impl StatusBarModule for BatteryModule {
    fn id(&self) -> &str {
        "battery"
    }

    fn render(&self, ctx: &ModuleContext) -> Vec<Segment> {
        let Some(bat) = &ctx.metrics.battery else {
            return vec![];
        };

        if !bat.present {
            return vec![];
        }

        let colors = battery_colors(ctx, bat.level);
        let icon = battery_icon(bat.level, bat.charging);
        let charging_indicator = if bat.charging { "\u{26a1}" } else { "" };

        let text = format!(" {} {}%{} ", icon, bat.level, charging_indicator);

        vec![Segment::new(&text, &colors.fg, &colors.bg)]
    }
}
